File: Medium/129 - SumRootToLeafNumbers/solution.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

/**
 * 执行用时：84 ms, 在所有 typescript 提交中击败了 82.35% 的用户
 * 内存消耗：40.6 MB, 在所有 typescript 提交中击败了 5.88% 的用户
 *
 * @param {(TreeNode | null)} root
 * @returns {number}
 */
function sumNumbers(root: TreeNode | null): number {
  if (!root) {
    return 0;
  }

  let sum: number = 0;

  let nodeQueue: TreeNode[] = [root];
  let valueQueue: number[] = [root.val];

  while (nodeQueue.length && valueQueue.length) {
    const currentNode: TreeNode = nodeQueue.shift() as TreeNode;
    let currentValue: number = valueQueue.shift() as number;

    if (!currentNode.left && !currentNode.right) {
      sum += currentValue;
    } else {
      if (currentNode.left) {
        nodeQueue.push(currentNode.left);
        valueQueue.push(currentValue * 10 + currentNode.left.val);
      }

      if (currentNode.right) {
        nodeQueue.push(currentNode.right);
        valueQueue.push(currentValue * 10 + currentNode.right.val);
      }
    }
  }

  return sum;
}

File: Medium/129 - SumRootToLeafNumbers/Solution.cs
// Question 129. Sum Root to Leaf Numbers (https://leetcode-cn.com/problems/sum-root-to-leaf-numbers/)

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class TreeNode
{
    public int val;
    public TreeNode left;
    public TreeNode right;
    public TreeNode(int x) { val = x; }
}

/// <summary>
/// 执行用时：112 ms, 在所有 C# 提交中击败了 58.82% 的用户
/// 内存消耗：24.3 MB, 在所有 C# 提交中击败了 16.33% 的用户
/// </summary>
public class Solution
{
    private int _sum = 0;

    public int SumNumbers(TreeNode root)
    {
        if (root == null) { return this._sum; }

        if (root.left == null && root.right == null) { return root.val; }

        this.TraverseUntilLeaf(root.left, root.val);
        this.TraverseUntilLeaf(root.right, root.val);

        return this._sum;
    }

    private void TraverseUntilLeaf(TreeNode root, int accumulatedValue)
    {
        if (root == null)
        {
            return;
        }

        accumulatedValue = accumulatedValue * 10 + root.val;

        Console.WriteLine($"root = {root.val}, accumulatedValue = {accumulatedValue}");

        if (root.left == null && root.right == null)
        {
            this._sum += accumulatedValue;
            return;
        }

        if (root.left != null)
        {
            this.TraverseUntilLeaf(root.left, accumulatedValue);
        }

        if (root.right != null)
        {
            this.TraverseUntilLeaf(root.right, accumulatedValue);
        }
    }
}